File: Problems/Algorithms/149. Max Points on a Line/max_points.py
#!/usr/bin/env python3

import sys
import math


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        if n <= 2:
            return n

        ans = 1
        for i in range(n-1):
            curr = 1
            lines = {}
            for j in range(i+1, n):
                x1, y1 = points[i][0], points[i][1]
                x2, y2 = points[j][0], points[j][1]
                slope = self.slope(x1, y1, x2, y2)
                lines[slope] = lines.get(slope, 1) + 1
                curr = max(lines[slope], curr)

            ans = max(ans, curr)

        return ans

    @staticmethod
    def slope(x1, y1, x2, y2):
        if x1 == x2:
            return (0, 0)
        if y1 == y2:
            return (sys.maxsize, sys.maxsize)

        delta_x, delta_y = x1 - x2, y1 - y2
        if delta_x < 0:
            delta_x, delta_y = -delta_x, -delta_y

        gcd = math.gcd(delta_x, delta_y)
        return (delta_x//gcd, delta_y//gcd)

File: Problems/Algorithms/149. Max Points on a Line/MaxPoints.java
class Solution {
    public int maxPoints(int[][] points) {
        int n = points.length;
        if (n <= 2) return n;

        int ans = 1;
        for (int i = 0; i < n-1; i++) {
            int curr = 1;
            HashMap<List<Integer>, Integer> lines = new HashMap<>();
            int x1 = points[i][0];
            int y1 = points[i][1];
            for (int j = i + 1; j < n; j++) {
                int x2 = points[j][0];
                int y2 = points[j][1];
                List<Integer> slope = Arrays.asList(getSlope(x1, y1, x2, y2));
                System.out.println(lines.get(slope));
                lines.put(slope, lines.getOrDefault(slope, 1) + 1);
                curr = Math.max(lines.get(slope), curr);
            }

            ans = Math.max(ans, curr);
        }

        return ans;
    }

    private Integer[] getSlope(int x1, int y1, int x2, int y2) {
        Integer[] slope = new Integer[2];

        if (x1 == x2) {
            slope[0] = 0;
            slope[1] = 0;
            return slope;
        }
        if (y1 == y2) {
            slope[0] = Integer.MAX_VALUE;
            slope[1] = Integer.MAX_VALUE;
            return slope;
        }

        int deltaX = x1 - x2;
        int deltaY = y1 - y2;
        if (deltaX < 0){
            deltaX = - deltaX;
            deltaY = - deltaY;
        }

        int gcd = getGCD(deltaX, deltaY);
        slope[0] = deltaX / gcd;
        slope[1] = deltaY / gcd;
        return slope;
    }

    private int getGCD(int a, int b) {
        if (b == 0) return a;
        return getGCD(b, a % b);
    }
}
